KEP-3488: Promote ValidatingAdmissionPolicy to Beta

alexzielenski opened the pull request:
#### What type of PR is this?


/kind feature

#### What this PR does / why we need it:

Prerequisites:
- https://github.com/kubernetes/kubernetes/pull/118642
- https://github.com/kubernetes/kubernetes/pull/119215

Graduation Criteria from KEP:

- [ ] Add tests which registers a validation policy for everything and iterates through all groups/versions/resources/subresources and ensures they get intercepted and work properly with a CEL validation policy
    - [x] Context: https://github.com/kubernetes/kubernetes/pull/113314#discussion_r1013596456
    - [ ] Implementation: https://github.com/kubernetes/kubernetes/pull/119409
- [x] Set paramKind in a ValidatingAdmissionPolicy results in starting a new informer that watches all instances of that object using a new unstructured informer which is inefficient
    - [x] Context: https://github.com/kubernetes/kubernetes/pull/113314#discussion_r1013331221
    - [x] Implementation: https://github.com/kubernetes/kubernetes/pull/115145
- [x] Switch to a lock-free implementation to address lock having to wait for all existing admission evaluations to complete and blocking all new admission evaluations until this completes.
    - [x] Context: https://github.com/kubernetes/kubernetes/pull/113314#discussion_r1013318103,https://github.com/kubernetes/kubernetes/pull/113314#discussion_r1013305167
    - [x] Implementation: https://github.com/kubernetes/kubernetes/pull/114527
- [x] Perform minimal possible number of conversions when evaluating multiple admission policies for a request resource. If multiple admission policies require the same conversion, convert only once. 
    - [x] Context: @liggitt: "webhook code loops up one level, first accumulates all the validation webhooks we'll run, then converts to the versions needed by those webhooks then evaluates in parallel"
    - [x] Partial Implementation: https://github.com/kubernetes/kubernetes/pull/115412
added caching of the conversion on per-policy basis, but multiple policies sharing paramKind may still reconvert
- [x] Authz check to the specific resource referenced in the policy's paramKind.
    - [x] Context: https://github.com/kubernetes/kubernetes/pull/113314#discussion_r1013135860
    - [x] Implementation: https://github.com/kubernetes/kubernetes/pull/113314/commits/d3f48136d05ba58550744233b60caa4b8d71012f
- [x] Complete feature of access to namespace metadata
    - [x] Implementation: https://github.com/kubernetes/kubernetes/pull/118267
- [ ] Complete type check for CRD: Purely advisory and should not block
    - [ ] Implementation: https://github.com/kubernetes/kubernetes/pull/119109
- [x] add controlled rollout strategy to support future CEL library/function/variable changes
    - [x] Implementation: https://github.com/kubernetes/kubernetes/pull/116779
- [x] [Quantity](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go#L100) support from CEL expression and tested properly
    - [x] CEL Library Implementation & tests: https://github.com/kubernetes/kubernetes/pull/118803
- [x] Support phase 2 features:
    - [x] [Informational type checking](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#informational-type-checking): https://github.com/kubernetes/kubernetes/pull/117377
    - [x] [Enforcement Actions](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#enforcement-actions): https://github.com/kubernetes/kubernetes/pull/115973
    - [x] [Audit Annotations](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#audit-annotations): https://github.com/kubernetes/kubernetes/pull/115973
    - [x] [Audit Events](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#audit-events)
    - [x] [Per namespace policy params](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#per-namespace-policy-params): https://github.com/kubernetes/kubernetes/pull/119215
    - [x] [Match Conditions](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#match-conditions): https://github.com/kubernetes/kubernetes/issues/116350
    - [x] [CEL Expression Composition](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#cel-expression-composition): https://github.com/kubernetes/kubernetes/issues/118642
    - [x] [Secondary Authz](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#secondary-authz): https://github.com/kubernetes/kubernetes/pull/116054
    - [x] [Access to namespace](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#access-to-namespace): https://github.com/kubernetes/kubernetes/pull/118267
    - [x] [Transition rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#transition-rules)
    - [x] [Resource constraints](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#resource-constraints)
    - [x] [Safety Features](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#safety-features)
    - [x] [Aggregated API servers](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#aggregated-api-servers)
    - [x] [CEL function library](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#cel-function-library)
        - [x] https://github.com/kubernetes/kubernetes/pull/108312
        - [x] https://github.com/kubernetes/kubernetes/pull/118803
        - [x] https://github.com/google/cel-go/pull/617
    - [x] [Client visibility](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#client-visibility)
    - [x] [Metrics](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#metrics): https://github.com/kubernetes/kubernetes/pull/112994

#### Which issue(s) this PR fixes:


#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Promoted API groups `ValidatingAdmissionPolicy` and `ValidatingAdmissionPolicyBinding` to `v1beta1`. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3488
```

/sig api-machinery
/cc @cici37 @jpbetz

jpbetz commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -458,6 +469,14 @@ type ValidatingAdmissionPolicyBindingSpec struct {
 	// Required.
 	ValidationActions []ValidationAction
+
+	NamespaceParamRef *NameSpaceParamRef
+}
+
+type NameSpaceParamRef struct {
+	FailAction *ParamFailureActionType
+	Name       string
+	Selector   *metav1.LabelSelector
```
+API docs on these fields

Some critical things to explain:

- Must set either Name or Selector but not both
- How FailAction works for both cases
- What happens with a Selector matches 0 param objects and when it matches more than one param objects

alexzielenski replied:
added API docs

jpbetz commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -458,6 +469,14 @@ type ValidatingAdmissionPolicyBindingSpec struct {
+	NamespaceParamRef *NameSpaceParamRef
+}
+
+type NameSpaceParamRef struct {
```
```suggestion
type NamespaceParamRef struct {
```

jpbetz commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -458,6 +469,14 @@ type ValidatingAdmissionPolicyBindingSpec struct {
+}
+
+type NameSpaceParamRef struct {
+	FailAction *ParamFailureActionType
```
If I remember right, the semantics of this field were not entirely obvious.  Is is the action that happens if the ref doesn't match anything?

The two cases are?

- Name is non-empty but refers to a non-existent object. This seems like a failure mode, so I agree "FailAction" is a sane name.
- Selector is non-empty, but doesn't select any objects.  This is a special condition. I'd like to support a fallback, but it might be non-obvious to users that this is the condition that triggers a "FailAction".

alexzielenski replied:
added API docs. Let me know if I you see anything I missed

jpbetz reviewed (changes requested):
@alexzielenski This looks ready for API review to me.  Would you set that up?

alexzielenski commented:
/label api-review

/cc @deads2k 
for API review

Still adding tests and tuning implementation. Expecting no further changes to types/global API code.

alexzielenski commented:
/triage accepted

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1beta1/types.go:
```diff
@@ -576,31 +592,575 @@ type ServiceReference struct {
+	// However, in order to prevent clusters from being put into an unstable state that cannot be recovered from via the API
+	// ValidatingAdmissionPolicy cannot match ValidatingAdmissionPolicy and ValidatingAdmissionPolicyBinding.
+	// Required.
+	MatchConstraints *MatchResources `json:"matchConstraints,omitempty" protobuf:"bytes,2,rep,name=matchConstraints"`
```
If this is required, why is it a pointer?

alexzielenski replied:
changed to a value

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1beta1/types.go:
```diff
@@ -576,31 +592,575 @@ type ServiceReference struct {
+
+// Supported binding param kinds.
+const (
+	BindingParamKindGlobal    BindingParamKind = "Global"
+	BindingParamKindNamespace BindingParamKind = "Namespace"
```
`Global` (maybe `ClusterWide`) and `PerNamespace`?  Only calling out the distinction since it's possible to use `ParamRef` to reference a parameter from one namespace and apply it to all namespaces.

alexzielenski replied:
renamed as suggested

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1beta1/types.go:
```diff
@@ -576,31 +592,575 @@ type ServiceReference struct {
+
+// ParamRef references a parameter resource
+// +structType=atomic
+type ParamRef struct {
```
Update type to match.

alexzielenski replied:
changed this type to `ClusterWideParamRef`

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1beta1/types.go:
```diff
@@ -576,31 +592,575 @@ type ServiceReference struct {
+	// +unionDiscriminator
+	Kind BindingParamKind `json:"kind" protobuf:"bytes,1,opt,name=kind"`
+
+	// GlobalParamRef specifies the parameter resource used to configure the
+	// admission control policy.
+	// If the resource referred to by ParamRef does not exist, this binding is
+	// considered mis-configured and the FailurePolicy of the
+	// ValidatingAdmissionPolicy applied.
```
```suggestion
	// GlobalParamRef specifies the parameter resource used to configure the
	// admission control policy used for cluster-scoped and all namespace-scoped requests.
	//
	// If the policy specifies a ParamKind and the the resource referred to by ParamRef does not exist, this binding is
	// considered mis-configured and the FailurePolicy of the
	// ValidatingAdmissionPolicy applied.
```
?

alexzielenski replied:
used suggestion as-is

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1beta1/types.go:
```diff
@@ -576,31 +592,575 @@ type ServiceReference struct {
+// NamespaceParamRef references a parameter resource scoped to the same
+// namespace as the resource being evaluated by a policy.
+// +structType=atomic
+type NamespaceParamRef struct {
```
can we actually AND the name and label selector?  Thinking ahead to a future with field selector for CRs, I can imagine adding field selectors and I think that ANDing the output of all selection criteria is the easiest to reason about.

alexzielenski replied:
added documentation to type for this detail, and removed note that name and selector are a union

deads2k commented on pkg/apis/admissionregistration/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,100 @@
+
+	dst.ParamRef = nil
+
+	// Restore the backup of namespaceParamRef annotation then wipe it
```
I don't understand this comment.

alexzielenski replied:
was vestigial. removed

deads2k commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -1086,13 +1086,77 @@ func validateValidatingAdmissionPolicyBindingSpec(spec *admissionregistration.Va
+	switch pr.Kind {
+	case "":
+		allErrors = append(allErrors, field.Required(fldPath.Child("kind"), ""))
+	case admissionregistration.BindingParamKindGlobal:
```
need to ensure that `NamespaceParamRef` is nil

alexzielenski replied:
added checking for this condition

deads2k commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -1086,13 +1086,77 @@ func validateValidatingAdmissionPolicyBindingSpec(spec *admissionregistration.Va
+		} else {
+			allErrors = append(allErrors, validateParamRef(pr.GlobalParamRef, fldPath.Child("globalParamRef"))...)
+		}
+	case admissionregistration.BindingParamKindNamespace:
```
need to ensure that `GlobalParamRef` is nil

alexzielenski replied:
added checking for this condition

deads2k commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -1086,13 +1086,77 @@ func validateValidatingAdmissionPolicyBindingSpec(spec *admissionregistration.Va
+	if npr.Selector != nil {
+		numParams += 1
+		labelSelectorValidationOpts := metav1validation.LabelSelectorValidationOptions{}
+		allErrors = append(allErrors, metav1validation.ValidateLabelSelector(npr.Selector, labelSelectorValidationOpts, fldPath)...)
```
fldPath should be the selector

alexzielenski replied:
updated to use correct fieldpath

deads2k commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -1086,13 +1086,77 @@ func validateValidatingAdmissionPolicyBindingSpec(spec *admissionregistration.Va
+	}
+
+	if numParams == 0 {
+		allErrors = append(allErrors, field.Required(fldPath, "must specify either name or selector"))
```
Based on the doc, I expected empty to mean "select all in the namespace"

alexzielenski replied:
Added doc and removed error

deads2k commented on pkg/apis/admissionregistration/v1beta1/defaults.go:
```diff
@@ -27,6 +27,30 @@ func addDefaultingFuncs(scheme *runtime.Scheme) error {
 	return RegisterDefaults(scheme)
 }
 
+// SetDefaults_ValidatingAdmissionPolicySpec sets defaults for ValidatingAdmissionPolicySpec
```
missing  default for a namespaceParameterRef FailAction

alexzielenski replied:
Added defaulting and doc

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -419,6 +431,59 @@ type ValidatingAdmissionPolicyBindingSpec struct {
+	// Allowed values are `Allow` or `Deny`
+	// Default to `Allow`
+	// +optional
+	FailAction *ParamFailureActionType `json:"failAction" protobuf:"bytes,3,opt,name=failAction,casttype=ParamFailureActionType"`
```
`ParameterNotFoundAction`?  The selection itself was valid.

alexzielenski replied:
Should the field itself also be renamed by the same logic?

deads2k replied:
> Should the field itself also be renamed by the same logic?

Yes and the json tag and godoc.

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1beta1/types.go:
```diff
@@ -576,31 +592,575 @@ type ServiceReference struct {
+	// Allowed values are `Allow` or `Deny`
+	// Default to `Allow`
+	// +optional
+	FailAction *ParamFailureActionType `json:"failAction,omitempty" protobuf:"bytes,3,opt,name=failAction,casttype=ParamFailureActionType"`
```
ParameterNotFoundAction? The selection itself was valid.

deads2k reviewed (changes requested):
I did not finish reviewing all the controller changes.  Can we separate the "api, defaulting, conversion, validation" from the controller and admission changes? I know there are behavioral changes required, but freeing up the API first broadens reviewers.

benluddy commented on pkg/registry/admissionregistration/validatingadmissionpolicybinding/authz.go:
```diff
@@ -85,26 +85,31 @@ func (v *validatingAdmissionPolicyBindingStrategy) authorize(ctx context.Context
+	if paramRef := binding.Spec.ParamRef; paramRef.Kind == admissionregistration.BindingParamKindGlobal && paramRef.GlobalParamRef != nil {
+		globalParamRef := paramRef.GlobalParamRef
+
+		//!TODO: can't perform static check on the namespaceParamRef
+		// should we copy this to the admission controller?
```
Even for bindings with cluster-scoped params, don't we want an authz check at point-of-use? Is there an issue to track this?

alexzielenski replied:
created

https://github.com/kubernetes/kubernetes/issues/119112

deads2k replied:
Can we simply require "list across all namespaces" permissions for the namespaceParamRef instead?  That seem very reasonable to require given the potential impact.

deads2k replied:
Because this is in the REST handler, it needs to be addressed when the API merges, we can't put it off.  I think 

```go
		attrs := authorizer.AttributesRecord{
			User:            user,
			Verb:            "list",
			ResourceRequest: true,
			Name:            "", // list requests don't have names
			Namespace:       "", // no namespace indicates list across all namespaces
			APIGroup:        apiGroup,
			APIVersion:      apiVersion,
			Resource:        resource,
```

is what you're looking for.

alexzielenski replied:
added in #119215 and moved internal changes to that PR. We can move discussion there if there are outstanding issues

alexzielenski commented on pkg/printers/internalversion/printers.go:
```diff
@@ -1647,10 +1647,18 @@ func printValidatingAdmissionPolicyBinding(obj *admissionregistration.Validating
+			}
+		} else if obj.Spec.ParamRef.Kind == admissionregistration.BindingParamKindNamespace {
+			if obj.Spec.ParamRef.NamespaceParamRef.Name != "" {
+				paramName = "*/" + obj.Spec.ParamRef.GlobalParamRef.Name
```
updated, thanks

cici37 commented:
Thank you for the PR! I haven't looked at the details but I would suggest to make `Adds NamespaceParamRef field` a separate PR from promoting the API to beta because we have other API changes pending(e.g. https://github.com/kubernetes/kubernetes/pull/118642). It might be good to promote the API to beta after all API changes merged. :)

alexzielenski committed:
test: adjust tests for v1beta1 validatingadmissionpolicy

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1beta1/types.go:
```diff
@@ -576,31 +592,587 @@ type ServiceReference struct {
+	// `kind` indicates which one of the other fields is non-empty.
+	// Required
+	// +unionDiscriminator
+	Kind BindingParamKind `json:"kind" protobuf:"bytes,1,opt,name=kind"`
```
```suggestion
	ParamKind BindingParamKind `json:"paramKind" protobuf:"bytes,1,opt,name=paramkind"`
```

and the godoc and other versions.  Don't want to confuse users with `kind` like "Pod".

alexzielenski replied:
renamed as suggested

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1beta1/types.go:
```diff
@@ -576,31 +592,587 @@ type ServiceReference struct {
+	// Allowed values are `Allow` or `Deny`
+	// Default to `Allow`
+	// +optional
+	FailAction *ParamFailureActionType `json:"failAction,omitempty" protobuf:"bytes,3,opt,name=failAction,casttype=ParamFailureActionType"`
```
```suggestion
	ParameterNotFoundAction *ParameterNotFoundActionType `json:"parameterNotFoundAction,omitempty" protobuf:"bytes,3,opt,name=parameterNotFoundAction,casttype=ParameterNotFoundActionType"`
```

and other versions.

The selection itself was valid, but there were no matches.

alexzielenski replied:
renamed as suggested

deads2k commented on test/integration/etcd/data.go:
```diff
@@ -358,14 +358,27 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
 		gvr("admissionregistration.k8s.io", "v1alpha1", "validatingadmissionpolicies"): {
 			Stub:             `{"metadata":{"name":"vap1","creationTimestamp":null},"spec":{"paramKind":{"apiVersion":"test.example.com/v1","kind":"Example"},"matchConstraints":{"resourceRules": [{"resourceNames": ["fakeName"], "apiGroups":["apps"],"apiVersions":["v1"],"operations":["CREATE", "UPDATE"], "resources":["deployments"]}]},"validations":[{"expression":"object.spec.replicas <= params.maxReplicas ...
 			ExpectedEtcdPath: "/registry/validatingadmissionpolicies/vap1",
+			ExpectedGVK:      gvkP("admissionregistration.k8s.io", "v1beta1", "ValidatingAdmissionPolicy"),
```
this looks off.  We should still be storing in v1alpha1 so that during an upgrade older servers can read the content written by newer servers.  I bet its controlled by the order of the versions added to the apiGroupInfo

alexzielenski replied:
Changed the default version for the types and the test to expect v1beta1 input to have v1alpha1 storage

alexzielenski commented:
Incorporated most of @deads2k feedback and rebased off of #119215 per @cici37's recommendation.

Will hold on this PR until #118642 and #119215 merge.

alexzielenski commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -1086,14 +1083,76 @@ func validateValidatingAdmissionPolicyBindingSpec(spec *admissionregistration.Va
+	return allErrors
+}
+
+func validateParamRef(pr *admissionregistration.ClusterWideParamRef, fldPath *field.Path) field.ErrorList {
```
renamed as suggested

alexzielenski commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -1086,14 +1083,76 @@ func validateValidatingAdmissionPolicyBindingSpec(spec *admissionregistration.Va
+		if pr.PerNamespace == nil {
+			allErrors = append(allErrors, field.Required(fldPath.Child("perNamespace"), fmt.Sprintf("required if Kind is `%s`", admissionregistration.BindingParamKindPerNamespace)))
+		} else {
+			allErrors = append(allErrors, validateNamespaceParamRef(pr.PerNamespace, fldPath.Child("perNamespae"))...)
```
renamed as suggested

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1beta1/types.go:
```diff
@@ -576,31 +592,587 @@ type ServiceReference struct {
+	// `kind` indicates which one of the other fields is non-empty.
+	// Required
+	// +unionDiscriminator
+	ParamKind BindingParamKind `json:"kind" protobuf:"bytes,1,opt,name=kind"`
```
```suggestion
	ParamKind BindingParamKind `json:"paramKind" protobuf:"bytes,1,opt,name=paramKind"`
```

missed the json tags and protobuf name

deads2k reviewed (changes requested):
missed  JSON tag and the authz check must exist in the REST handler.

Atharva-Shinde commented:
[exception request received](https://groups.google.com/a/kubernetes.io/g/release-team/c/NtaSED_F1a8/m/2JYiKx2MAwAJ).
Holding for it's approval.
/hold

alexzielenski committed:
api: add v1beta1 ValidatingAdmissionPolicy/Binding types and storage

alexzielenski commented:
Update for reviewers: The exception has been approved [here](https://groups.google.com/g/kubernetes-sig-release/c/4yBLh6u8daw).
/hold cancel

(needs milestone?)

cici37 commented:
/milestone v1.28

cici37 commented:
@alexzielenski Could you also update the feature gate in this PR? Under kube_features.go, make the release phrase as `Beta` but keep default as `false`. e.g. `ValidatingAdmissionPolicy: {Default: false, PreRelease: featuregate.Beta},` Thank you

cici37 commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -1151,9 +1151,34 @@ func validateParamRef(pr *admissionregistration.ParamRef, fldPath *field.Path) f
+		}
+	}
+
+	if len(pr.ParameterNotFoundAction) > 0 {
```
Check if ParameterNotFoundAction is set and return required err if not. Add test accordingly as well.

alexzielenski replied:
The current implementation returns a `field.Required` error if `len(ParameterNotFoundAction)` == 0. Test is here:

https://github.com/kubernetes/kubernetes/blob/2a809d9011b330304fee9b6a380a3a7ea11d20b8/pkg/apis/admissionregistration/validation/validation_test.go#L3906-L3921

cici37 reviewed (commented):
Things might also needed for API promotion:
- Update all implementation to switch from using `k8s.io/api/admissionregistration/v1alpha1` to `k8s.io/api/admissionregistration/v1beta1`
- Update the integration w/e2e test to use beta1 instead of alpha1.
- Update feature gate to beta and still turn off by default.

Thank you :)

alexzielenski committed:
refactor: replace usage of v1alpha1 with v1beta1

v1alpha -> v1beta

fill in DenyAction where there is no ParameterNotFoundAction

alexzielenski committed:
update codegen

alexzielenski commented:
@cici37 I've implemented the changes you've suggested. Ready for another pass whenever you have time.

cici37 commented:
Worth to edit the PR description to state all beta graduation criteria be addressed? 
https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3488-cel-admission-control/README.md#beta

cici37 commented:
/api-review
This should be ready for API review since https://github.com/kubernetes/kubernetes/pull/119215 has been approved. 
Currently rebased on #119215 , the changes will be more clear after #119215 merged.

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go:
```diff
@@ -390,31 +390,42 @@ func (c *celAdmissionController) Validate(
+				message = fmt.Sprintf("ValidatingAdmissionPolicy '%s' denied request: %s", decision.Definition.Name, decision.Message)
+			}
+
+			//!TODO: does it make sense to use decision.Reason for Cause.Type?
```
I know we can't use it directly since it is a different type, just seems odd that every rule provides a `Reason` but we only use it first rule we happen to evaluate.

cici37 replied:
I didn't expect code change in this PR.
We didn't provide the cause type since we didn't expect user to check the type based on what is inside details.Causes. I agree it would be a good candidate for further improvement.

alexzielenski replied:
discussed offline that we will keep the bug fix but remove todo

cici37 commented:
/retest

deads2k commented on pkg/features/kube_features.go:
```diff
@@ -1197,7 +1197,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	genericfeatures.APIResponseCompression: {Default: true, PreRelease: featuregate.Beta},
 
-	genericfeatures.ValidatingAdmissionPolicy: {Default: false, PreRelease: featuregate.Alpha},
+	genericfeatures.ValidatingAdmissionPolicy: {Default: false, PreRelease: featuregate.Beta},
```
some items on the checklist are still open.  Unless the PR solves all of them you may be better off to  leave this one line out.

cici37 replied:
Captured the work for GA here: https://github.com/kubernetes/enhancements/pull/4132
Thank you!

deads2k commented on pkg/registry/admissionregistration/rest/storage_apiserver.go:
```diff
@@ -116,6 +123,40 @@ func (p RESTStorageProvider) v1alpha1Storage(apiResourceConfigSource serverstora
+		if err != nil {
+			return storage, err
+		}
+		policyGetter = policyStorage
```
While this works, I'd say that my life got better in openshift when we switched most usages to clients instead of storage links.  Not requiring a change now, but noting that additional hop through the main API has some advantages related to auditing

deads2k commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/matching/matching_test.go:
```diff
@@ -78,12 +78,12 @@ func TestMatcher(t *testing.T) {
 	mapper.RegisterKindFor(gvr("apps", "v1", "deployments"), "", gvk("apps", "v1", "Deployment"))
 	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "", gvk("apps", "v1beta1", "Deployment"))
-	mapper.RegisterKindFor(gvr("apps", "v1alpha1", "deployments"), "", gvk("apps", "v1alpha1", "Deployment"))
+	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "", gvk("apps", "v1beta1", "Deployment"))
```
why did this have to change?  find/replace perhaps?  Same a few lines down.

deads2k replied:
Yeah this should be restored.

deads2k commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/matching/matching_test.go:
```diff
@@ -78,12 +78,12 @@ func TestMatcher(t *testing.T) {
 	mapper.RegisterKindFor(gvr("apps", "v1", "deployments"), "scale", gvk("autoscaling", "v1", "Scale"))
 	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "scale", gvk("apps", "v1beta1", "Scale"))
-	mapper.RegisterKindFor(gvr("apps", "v1alpha1", "deployments"), "scale", gvk("apps", "v1alpha1", "Scale"))
+	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "scale", gvk("apps", "v1beta1", "Scale"))
```
and don't change this line either.

deads2k commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/matching/matching_test.go:
```diff
@@ -673,12 +673,12 @@ func BenchmarkShouldCallHookWithComplexRule(b *testing.B) {
 	mapper.RegisterKindFor(gvr("apps", "v1", "deployments"), "", gvk("apps", "v1", "Deployment"))
 	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "", gvk("apps", "v1beta1", "Deployment"))
-	mapper.RegisterKindFor(gvr("apps", "v1alpha1", "deployments"), "", gvk("apps", "v1alpha1", "Deployment"))
+	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "", gvk("apps", "v1beta1", "Deployment"))
```
accidental change

deads2k commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/matching/matching_test.go:
```diff
@@ -673,12 +673,12 @@ func BenchmarkShouldCallHookWithComplexRule(b *testing.B) {
 	mapper.RegisterKindFor(gvr("apps", "v1", "deployments"), "scale", gvk("autoscaling", "v1", "Scale"))
 	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "scale", gvk("apps", "v1beta1", "Scale"))
-	mapper.RegisterKindFor(gvr("apps", "v1alpha1", "deployments"), "scale", gvk("apps", "v1alpha1", "Scale"))
+	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "scale", gvk("apps", "v1beta1", "Scale"))
```
accidental change

deads2k commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/matching/matching_test.go:
```diff
@@ -743,12 +743,12 @@ func BenchmarkShouldCallHookWithComplexSelectorAndRule(b *testing.B) {
 	mapper.RegisterKindFor(gvr("apps", "v1", "deployments"), "", gvk("apps", "v1", "Deployment"))
 	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "", gvk("apps", "v1beta1", "Deployment"))
-	mapper.RegisterKindFor(gvr("apps", "v1alpha1", "deployments"), "", gvk("apps", "v1alpha1", "Deployment"))
+	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "", gvk("apps", "v1beta1", "Deployment"))
```
accidental change

deads2k commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/matching/matching_test.go:
```diff
@@ -743,12 +743,12 @@ func BenchmarkShouldCallHookWithComplexSelectorAndRule(b *testing.B) {
 	mapper.RegisterKindFor(gvr("apps", "v1", "deployments"), "scale", gvk("autoscaling", "v1", "Scale"))
 	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "scale", gvk("apps", "v1beta1", "Scale"))
-	mapper.RegisterKindFor(gvr("apps", "v1alpha1", "deployments"), "scale", gvk("apps", "v1alpha1", "Scale"))
+	mapper.RegisterKindFor(gvr("apps", "v1beta1", "deployments"), "scale", gvk("apps", "v1beta1", "Scale"))
```
accidental change

deads2k commented on staging/src/k8s.io/apiserver/pkg/features/kube_features.go:
```diff
@@ -241,7 +241,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	APIServerTracing: {Default: true, PreRelease: featuregate.Beta},
 
-	ValidatingAdmissionPolicy: {Default: false, PreRelease: featuregate.Alpha},
+	ValidatingAdmissionPolicy: {Default: false, PreRelease: featuregate.Beta},
```
same comment about whether this is going to fix the remaining items on the checklist.

alexzielenski replied:
reverted

deads2k commented:
API and impl changes look good. The tests need fixing before merge and the tests need to be added for the beta.

/approve

/assign @jpbetz @cici37 
for lgtm

cici37 commented:
Thank you! @alexzielenski Please update following when you got time. 
- Revert the feature gate change in this PR
- Fix the accidentally touch for version in match_test

Thank you! ^^

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go:
```diff
@@ -390,31 +390,41 @@ func (c *celAdmissionController) Validate(
-			message = fmt.Sprintf("ValidatingAdmissionPolicy '%s' with binding '%s' denied request: %s", deniedDecision.Definition.Name, deniedDecision.Binding.Name, deniedDecision.Message)
-		} else {
-			message = fmt.Sprintf("ValidatingAdmissionPolicy '%s' denied request: %s", deniedDecision.Definition.Name, deniedDecision.Message)
+		var causes []metav1.StatusCause
```
Since we only return the first decision back, constructing causes here looks redundant to me. Also returning causes for all validation as err.ErrStatus.Details.Causes does not sound right. Could we exclude this change from current PR?

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go:
```diff
@@ -390,31 +390,41 @@ func (c *celAdmissionController) Validate(
 		err.ErrStatus.Reason = reason
 		err.ErrStatus.Code = reasonToCode(reason)
-		err.ErrStatus.Details.Causes = append(err.ErrStatus.Details.Causes, metav1.StatusCause{Message: message})
+		err.ErrStatus.Details.Causes = causes
```
Does not sound right here. Remove it from current PR?

cici37 commented:
/lgtm
Thank you!

cici37 commented:
/retest

alexzielenski commented:
/retest

sftim commented:
The KEP link for the changelog should be to k/enhancements, not -  https://github.com/cici37/enhancements/

sftim commented:
Changelog suggestion:
```markdown
- Promoted validating admission policies to beta
- Promoted API groups `ValidatingAdmissionPolicy` and `ValidatingAdmissionPolicyBinding` to `v1beta1`. 
```
